### PR TITLE
Add read downsampling as a pre processing step

### PR DESCRIPTION
**Pull request by mhkc**
# Description
This PR adds an optional preprocessing step for downsampling the number of reads to a given target. This because extremely high sequencing depth can negatively affect the *de-novo* assembly quality.

## Primary function of PR
- [ ] Hotfix
- [x] Minor functionality improvement
- [ ] Major functionality improvement / New type of analysis
- [ ] Backward-breaking functionality

# Sign-offs
- [ ] Code reviewed by @octocat
- [ ] Code tested by @octocat

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added `get_taxon` to methods
 - Added `streptococcus` and `streptococcus_pyogenes` workflows and profiles to configs
 - Added ska filepath to yaml
+- Added optional read downsampling using seqtk.
```
Comment (ryanjameskennedy) [inline CHANGELOG.md:20]: We're not adding the removal of human reads right?

File: workflows/escherichia_coli.nf [hunk 2]
```diff
@@ -26,16 +26,20 @@ include { virulencefinder                           } from '../nextflow-modules/
...
         .map{ row -> get_meta(row) }
-        .branch {
-            iontorrent: it[2] == "iontorrent"
-            illumina: it[2] == "illumina"
-            nanopore: it[2] == "nanopore"
-        }
+        .map{ row -> [row[0], row[2]] }
         .set{ ch_meta }
 
+    // Create channel for reads
```
Comment (ryanjameskennedy) [inline workflows/escherichia_coli.nf:36]: Is there not a way to do this without having to read the csv file twice?

File: workflows/klebsiella_pneumoniae.nf [hunk 3]
```diff
@@ -25,16 +25,20 @@ include { virulencefinder                           } from '../nextflow-modules/
...
-            nanopore: it[2] == "nanopore"
-        }
+        .map{ row -> [row[0], row[2]] }
         .set{ ch_meta }
 
+    // Create channel for reads
+    Channel.fromPath(params.csv)
+        .splitCsv(header:true)
+        .map{ row -> get_meta(row) }
+        .map{ row -> [row[0], row[1]] }
```
Comment (ryanjameskennedy) [inline workflows/klebsiella_pneumoniae.nf:39]: Think you should rather create a method that returns the desired tuple and `.tap()` to output two channels from reading one

File: workflows/staphylococcus_aureus.nf [hunk 4]
```diff
@@ -24,16 +24,20 @@ include { virulencefinder                           } from '../nextflow-modules/
...
         .map{ row -> get_meta(row) }
-        .branch {
-            iontorrent: it[2] == "iontorrent"
-            illumina: it[2] == "illumina"
-            nanopore: it[2] == "nanopore"
-        }
+        .map{ row -> [row[0], row[2]] }
         .set{ ch_meta }
 
+    // Create channel for reads
```
Comment (ryanjameskennedy) [inline workflows/staphylococcus_aureus.nf:34]: See prev wf comment

**Review by ryanjameskennedy — CHANGES_REQUESTED** [review]
Just the few changes. You need to add seqtk download to the Makefile in the container dir. Otherwise looks good!

File: configs/nextflow.ngp.config [hunk 5]
```diff
@@ -290,6 +290,11 @@ process {
 		container = "${params.containerDir}/bonsai-prp.sif"
 		publishDir = [ path: "${params.outdir}/${params.speciesDir}/analysis_metadata", mode: 'copy', overwrite: true ]
 	}
+	withName: seqtk_sample {
+		container = "https://depot.galaxyproject.org/singularity/seqtk:1.4--he4a0461_2"
```
Comment (ryanjameskennedy) [inline configs/nextflow.ngp.config:294]:
```suggestion
		container = "${params.containerDir}/seqtk.sif"
```

File: configs/nextflow.base.config
Comment (ryanjameskennedy) [inline configs/nextflow.base.config:1]: Add `params.targetSampleSize = null` to profiles

File: CHANGELOG.md (hunk 1 above)
Comment (mhkc) [reply CHANGELOG.md:20]: Not in this one. Better to do feature by feature so we can ensure that everything works

File: workflows/escherichia_coli.nf (hunk 2 above)
Comment (mhkc) [reply workflows/escherichia_coli.nf:36]: Yes there are but I think that requires setting temporary variables.

File: nextflow-modules/modules/hostile/main.nf
Comment (ryanjameskennedy) [inline nextflow-modules/modules/hostile/main.nf:1]: This is for the human read removal right?

File: workflows/bacterial_base.nf [hunk 6]
```diff
@@ -22,37 +25,59 @@ workflow CALL_BACTERIAL_BASE {
...
+            .map{ row -> get_meta(row) }
+            .set{ ch_meta }
+
+        // Create channel for reads
+        ch_raw_input
+            .map{ row -> get_reads(row) }
+            .set{ ch_reads }
+
+        // downsample reads
+        seqtk_sample( ch_reads.map(row -> [row[0], row[1], params.targetSampleSize]) ).reads
```
Comment (ryanjameskennedy) [inline workflows/bacterial_base.nf:46]: Are we not separating `params.targetSampleSize` into it's own channel? We agreed that `params.targetSampleSize` is not misleading right?

File: nextflow-modules/modules/seqtk/main.nf [hunk 7]
```diff
@@ -0,0 +1,54 @@
+process seqtk_sample {
+  tag "${sampleID}"
+  scratch params.scratch
+
+  input:
+    tuple val(sampleID), path(reads), val(sample_size)
```
Comment (ryanjameskennedy) [inline nextflow-modules/modules/seqtk/main.nf:6]: See comment about separating channels

Comment (mhkc) [reply nextflow-modules/modules/seqtk/main.nf:6]: Yes but this was the best solution that causes the least amount of issues

Comment (mhkc) [reply nextflow-modules/modules/seqtk/main.nf:6]: Nextflow requires input channels to be defined. The `sample_size` parameter can be either an int, a float or a null which causes nextflow to crash if set as a separate channel but its allowed if passed with the rest of the data.

File: workflows/bacterial_base.nf (hunk 6 above)
Comment (mhkc) [reply workflows/bacterial_base.nf:58]: See other comment.

File: nextflow-modules/modules/seqtk/main.nf (hunk 7 above)
Comment (ryanjameskennedy) [reply nextflow-modules/modules/seqtk/main.nf:6]: Quast has a channel that can be null (`referenceGenome`) when using the `streptococcus` profile, can't we copy the same logic?

**Review by ryanjameskennedy — APPROVED** [review]
LGTM!

**Pull request merged**